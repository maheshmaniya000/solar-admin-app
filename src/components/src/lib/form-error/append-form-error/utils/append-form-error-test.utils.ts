import { ElementRef, Type } from '@angular/core';
import {
  AbstractControl,
  AsyncValidatorFn,
  ValidationErrors
} from '@angular/forms';
import { SpectatorDirective } from '@ngneat/spectator/jest';
import { isEmpty } from 'lodash-es';
import { Observable, of } from 'rxjs';
import { delay } from 'rxjs/operators';

import { AppendFormErrorDirective } from '../append-form-error.directive';

export class AppendFormErrorTestUtils {
  static readonly asyncRequired: (
    delayInMilliseconds: number
  ) => AsyncValidatorFn =
    (delayInMilliseconds: number) =>
    (control: AbstractControl): Observable<ValidationErrors> =>
      isEmpty(control.value)
        ? of({ required: true }).pipe(delay(delayInMilliseconds))
        : of(null);

  static expectDirectiveToBePresentOn<D extends AppendFormErrorDirective>(
    element: Element,
    type: Type<D>,
    spectator: SpectatorDirective<AppendFormErrorDirective>
  ): void {
    const appendFormErrorDirectiveElement: Element = spectator.query(type, {
      read: ElementRef
    }).nativeElement;
    expect(appendFormErrorDirectiveElement).toBe(element);
  }
}
