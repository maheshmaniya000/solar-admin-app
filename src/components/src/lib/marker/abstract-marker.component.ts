import {
  Component,
  ElementRef,
  EventEmitter,
  HostBinding,
  HostListener,
  Input,
  Output
} from '@angular/core';

import { Vector2 } from '@solargis/types/pvlib-api';

@Component({ template: '' })
export abstract class AbstractMarkerComponent {
  @HostBinding('class.selected') @Input() selected = false;
  @Output() selectedChange = new EventEmitter<void>();

  constructor(readonly elementRef: ElementRef<HTMLElement>) {}

  @HostListener('click') onClick(): void {
    this.selectedChange.emit();
  }

  getPivotPoint(): Vector2 {
    const element = this.elementRef.nativeElement;

    return {
      u: element.offsetWidth / 2,
      v: element.offsetHeight
    };
  }
}
