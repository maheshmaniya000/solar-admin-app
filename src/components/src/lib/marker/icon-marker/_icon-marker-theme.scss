@use '~@angular/material' as mat;
@use 'palette';
@use '../lib/icon/icon' as icon;

@mixin theme($theme-or-color-config) {
  $color-config: mat.get-color-config($theme-or-color-config);
  $accent: map-get($color-config, accent);

  sg-icon-marker {
    path {
      fill: palette.$dark-text;
    }

    &:hover {
      path {
        fill: black;
      }
    }

    &.selected {
      path {
        fill: mat.get-color-from-palette($accent);
        stroke: white;
      }

      &:hover {
        path {
          fill: mat.get-color-from-palette($accent, darker);
        }
      }
    }

    mat-icon.mat-icon {
      @include icon.size(inline);
      color: white;
    }
  }
}
