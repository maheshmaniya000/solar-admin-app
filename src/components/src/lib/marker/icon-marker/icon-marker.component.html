<svg
  xmlns="http://www.w3.org/2000/svg"
  xmlns:xlink="http://www.w3.org/1999/xlink"
  viewBox="0 0 22 33"
  [ngClass]="{ small: !isIconVisible() }"
>
  <defs>
    <style>
      .b {
        fill: url(#a);
      }
    </style>
    <radialGradient
      id="a"
      cx="0.5"
      cy="0.5"
      r="0.5"
      gradientUnits="objectBoundingBox"
    >
      <stop offset="0" stop-opacity="0.396" />
      <stop offset="1" stop-opacity="0" />
    </radialGradient>
  </defs>
  <ellipse class="b" cx="7" cy="3.5" rx="7" ry="3.5" />
  <path
    d="M10,28l7.6-11.4C19.5,13.737,20,12.255,20,10A10,10,0,0,0,0,10c0,2.255.511,3.753,2.4,6.6S10,28,10,28Z"
  />
</svg>
<mat-icon *ngIf="isIconVisible()">{{ icon }}</mat-icon>
