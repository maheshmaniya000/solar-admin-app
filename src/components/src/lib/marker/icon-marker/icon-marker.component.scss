:host {
  position: relative;
  display: inline-block;
  cursor: pointer;

  svg {
    width: 22px;
    height: 33px;

    path {
      stroke-linejoin: round;
      transform: translate(1px, 1px);
    }

    ellipse {
      transform: translate(4px, 26px);
    }

    &.small {
      width: 18px;
      height: 26px;

      path {
        transform: scale(0.952) translate(1.5px, 1.5px);
        stroke-width: 1.269px;
      }

      ellipse {
        transform: scale(1.269) translate(1.5px, 19.5px);
      }
    }
  }

  .mat-icon {
    position: absolute;
    top: 4px;
    left: 3px;
  }
}
