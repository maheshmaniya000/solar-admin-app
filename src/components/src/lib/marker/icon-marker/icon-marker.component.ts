import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { isEmpty } from 'lodash-es';

import { Vector2 } from '@solargis/types/pvlib-api';

import { AbstractMarkerComponent } from '../abstract-marker.component';

@Component({
  selector: 'sg-icon-marker',
  templateUrl: './icon-marker.component.html',
  styleUrls: ['./icon-marker.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
  providers: [
    // eslint-disable-next-line no-use-before-define
    { provide: AbstractMarkerComponent, useExisting: IconMarkerComponent }
  ]
})
export class IconMarkerComponent extends AbstractMarkerComponent {
  @Input() icon: string;

  isIconVisible(): boolean {
    return !isEmpty(this.icon);
  }

  getPivotPoint(): Vector2 {
    const pivotPoint = super.getPivotPoint();
    pivotPoint.v *= 0.75;
    return pivotPoint;
  }
}
