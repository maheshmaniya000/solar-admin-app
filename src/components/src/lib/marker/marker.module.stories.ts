import { MatButtonModule } from '@angular/material/button';
import { moduleMetadata, Story } from '@storybook/angular';

import { SgDefaultsModule } from '../defaults/defaults.module';
import { MarkerModule } from './marker.module';

export default {
  title: 'Solargis/Components/Marker',
  decorators: [
    moduleMetadata({
      imports: [MarkerModule, MatButtonModule, SgDefaultsModule]
    })
  ]
};

export const singleSelect: Story = props => ({
  props,
  template: `
    <div class="container" data-test="single-select-marker-component-story">
      <div class="markers-grid">
        <span class="label">Empty icon marker</span>
        <sg-icon-marker [selected]="selected"></sg-icon-marker>
        <span class="label">Icon marker</span>
        <sg-icon-marker icon="latlon_16px" [selected]="selected"></sg-icon-marker>
        <span class="label">Text marker</span>
        <sg-text-marker text="abcd" [selected]="selected"></sg-text-marker>
        <span class="label">Large text marker</span>
        <sg-text-marker class="large" text="ab" [selected]="selected"></sg-text-marker>
        <span class="label">Text marker without caret</span>
        <sg-text-marker text="abcd" [selected]="selected" [caretHidden]="true"></sg-text-marker>
        <span class="label">Large text marker without caret</span>
        <sg-text-marker class="large" text="ab" [selected]="selected" [caretHidden]="true"></sg-text-marker>
      </div>
    </div>
    `,
  styles: [
    `
    .container {
      height: 100vh;
      background-color: #a5adba;
    }

    .markers-grid {
      display: grid;
      width: 270px;
      grid-template-columns: repeat(2, auto);
      grid-template-rows: repeat(4, auto);
      grid-gap: 35px;
      padding: 15px;
    }

    .label {
      align-self: center;
    }
    `
  ]
});

singleSelect.argTypes = {
  selected: {
    name: 'Selected',
    control: { type: 'radio', options: [true, false] },
    defaultValue: false
  }
};
