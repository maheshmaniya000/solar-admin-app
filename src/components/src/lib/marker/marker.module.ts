import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { MatIconModule } from '@angular/material/icon';

import { IconMarkerComponent } from './icon-marker/icon-marker.component';
import { TextMarkerComponent } from './text-marker/text-marker.component';

@NgModule({
  declarations: [IconMarkerComponent, TextMarkerComponent],
  exports: [IconMarkerComponent, TextMarkerComponent],
  imports: [MatIconModule, CommonModule]
})
export class MarkerModule {}
