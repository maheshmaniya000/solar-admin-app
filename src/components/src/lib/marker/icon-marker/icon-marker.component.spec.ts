import { ElementRef } from '@angular/core';
import { createComponentFactory, Spectator } from '@ngneat/spectator/jest';

import { mockElementDimensions } from '../../utils/test';
import { MarkerModule } from '../marker.module';
import { IconMarkerComponent } from './icon-marker.component';

import spyOn = jest.spyOn;

describe('IconMarkerComponent', () => {
  describe('unit', () => {
    let component: IconMarkerComponent;

    beforeEach(() => {
      const element = document.createElement('div');
      mockElementDimensions(element, 20, 40);
      component = new IconMarkerComponent(new ElementRef(element));
    });

    describe('getPivotPoint', () => {
      it(`should return the icon marker's pivot point`, () => {
        component.icon = 'some-icon';
        expect(component.getPivotPoint()).toEqual({ u: 10, v: 30 });
      });
    });
  });

  describe('component', () => {
    let spectator: Spectator<IconMarkerComponent>;
    const createComponent = createComponentFactory({
      component: IconMarkerComponent,
      imports: [MarkerModule]
    });

    beforeEach(() => (spectator = createComponent()));

    it('should render empty marker with smaller svg', () => {
      expect(spectator.fixture).toMatchSnapshot();
    });

    it('should render marker with bigger svg and with icon', () => {
      spectator.setInput('icon', 'latlon_16px');
      expect(spectator.fixture).toMatchSnapshot();
    });

    it('should emit selectedChange event when clicked', () => {
      const emitSpy = spyOn(spectator.component.selectedChange, 'emit');
      spectator.click(spectator.element);
      expect(emitSpy).toHaveBeenCalledTimes(1);
      spectator.setInput('selected', true);
      spectator.click(spectator.element);
      expect(emitSpy).toHaveBeenCalledTimes(2);
    });

    it('should toggle selected class when selected property changes', () => {
      expect(spectator.element).not.toHaveClass('selected');
      spectator.setInput('selected', true);
      spectator.detectChanges();
      expect(spectator.element).toHaveClass('selected');
      spectator.setInput('selected', false);
      spectator.detectChanges();
      expect(spectator.element).not.toHaveClass('selected');
    });
  });
});
