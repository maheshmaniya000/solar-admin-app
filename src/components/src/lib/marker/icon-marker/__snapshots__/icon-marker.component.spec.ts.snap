// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconMarkerComponent component should render empty marker with smaller svg 1`] = `
<sg-icon-marker
  __ngContext__={[Function LRootView]}
  elementRef={[Function ElementRef]}
  selected="false"
  selectedChange={[Function EventEmitter_]}
>
  <svg
    class="small"
    viewBox="0 0 22 33"
    xmlns="http://www.w3.org/2000/svg"
    xmlns:xlink="http://www.w3.org/1999/xlink"
  >
    <defs>
      <style>
         .b { fill: url(#a); } 
      </style>
      <radialgradient
        cx="0.5"
        cy="0.5"
        gradientUnits="objectBoundingBox"
        id="a"
        r="0.5"
      >
        <stop
          offset="0"
          stop-opacity="0.396"
        />
        <stop
          offset="1"
          stop-opacity="0"
        />
      </radialgradient>
    </defs>
    <ellipse
      class="b"
      cx="7"
      cy="3.5"
      rx="7"
      ry="3.5"
    />
    <path
      d="M10,28l7.6-11.4C19.5,13.737,20,12.255,20,10A10,10,0,0,0,0,10c0,2.255.511,3.753,2.4,6.6S10,28,10,28Z"
    />
  </svg>
</sg-icon-marker>
`;

exports[`IconMarkerComponent component should render marker with bigger svg and with icon 1`] = `
<sg-icon-marker
  __ngContext__={[Function LRootView]}
  elementRef={[Function ElementRef]}
  icon={[Function String]}
  selected="false"
  selectedChange={[Function EventEmitter_]}
>
  <svg
    viewBox="0 0 22 33"
    xmlns="http://www.w3.org/2000/svg"
    xmlns:xlink="http://www.w3.org/1999/xlink"
  >
    <defs>
      <style>
         .b { fill: url(#a); } 
      </style>
      <radialgradient
        cx="0.5"
        cy="0.5"
        gradientUnits="objectBoundingBox"
        id="a"
        r="0.5"
      >
        <stop
          offset="0"
          stop-opacity="0.396"
        />
        <stop
          offset="1"
          stop-opacity="0"
        />
      </radialgradient>
    </defs>
    <ellipse
      class="b"
      cx="7"
      cy="3.5"
      rx="7"
      ry="3.5"
    />
    <path
      d="M10,28l7.6-11.4C19.5,13.737,20,12.255,20,10A10,10,0,0,0,0,10c0,2.255.511,3.753,2.4,6.6S10,28,10,28Z"
    />
  </svg><mat-icon
    aria-hidden="true"
    class="mat-icon notranslate material-icons mat-icon-no-color"
    data-mat-icon-type="font"
    role="img"
  >
    latlon_16px
  </mat-icon>
</sg-icon-marker>
`;
